### PR TITLE
cmake: fix GENERATE_REFERENCE_OUTPUT for failing tests

### LINKED ISSUE #5755: ctest GENERATE_REFERENCE_OUTPUT with failing tests
If a test crashes, ``GENERATE_REFERENCE_OUTPUT=1`` will report the test as succeeding (which is clearly wrong and confusing) and also overwrite the existing test output with the error message. We should fix that.

### PR DESCRIPTION
**Pull request by tjhei**
fixes #5755

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -315,18 +315,24 @@ FOREACH(_test ${_tests})
...
+    # We are creating an empty file "/runs" if the test runs successfully without
+    # crashing. This is used later to only diff if necessary.
     IF("${_mpi_count}" STREQUAL "1")
       ADD_CUSTOM_COMMAND(
         OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/screen-output
+  COMMAND rm -f ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/runs
         COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/cmake/timeout.sh ${TEST_TIME_LIMIT}s ${CMAKE_CURRENT_SOURCE_DIR}/cmake/run_test.sh ${EXPECT} ${ASPECT_BINARY} ${CMAKE_CURRENT_BINARY_DIR}/${_testname}.x.prm ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/screen-output.tmp
         COMMAND ${_replacement_script} screen-output <${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/screen-output.tmp >${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/screen-output
+  COMMAND touch ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/runs
         DEPENDS ${ASPECT_BINARY} ${CMAKE_CURRENT_BINARY_DIR}/${_testname}.x.prm ${_testdepends})
```
Comment (bangerth) [inline tests/CMakeLists.txt:327]: Can you indent consistently here?

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -315,18 +315,24 @@ FOREACH(_test ${_tests})
...
     ELSE()
       ADD_CUSTOM_COMMAND(
         OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/screen-output
+  COMMAND rm -f ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/runs
         COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/cmake/timeout.sh ${TEST_TIME_LIMIT}s mpirun -np ${_mpi_count} ${ASPECT_BINARY} ${CMAKE_CURRENT_BINARY_DIR}/${_testname}.x.prm
                 > ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/screen-output.tmp
         COMMAND ${_replacement_script} screen-output <${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/screen-output.tmp >${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/screen-output
+  COMMAND touch ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/runs
         DEPENDS ${ASPECT_BINARY} ${CMAKE_CURRENT_BINARY_DIR}/${_testname}.x.prm ${_testdepends})
```
Comment (bangerth) [inline tests/CMakeLists.txt:336]: And here

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -363,8 +369,10 @@ FOREACH(_test ${_tests})
         # the normalized output file:
         SET(_run_file ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/${_name}.notime)
 
-        # generate .diff file:
+        # generate .diff file, only run diff if "/runs" exists (test ran successfully),
+  # otherwise GENERATE_REFERENCE_OUTPUT would overwrite reference data with garbage.
         ADD_CUSTOM_COMMAND(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/${_name}.diff
+    COMMAND test -f ${CMAKE_CURRENT_BINARY_DIR}/output-${_testname}/runs
```
Comment (bangerth) [inline tests/CMakeLists.txt:375]: And here

**Comment by tjhei** [comment]
fixed

**Review by bangerth — APPROVED** [review]

**Pull request merged**